packages/datadog-instrumentations/src/mocha.js

```diff
@@ -5,101 +5,132 @@ const testStartCh = channel('ci:mocha:test:start')
 const errorCh = channel('ci:mocha:test:error')
 const skipCh = channel('ci:mocha:test:skip')
 const testFinishCh = channel('ci:mocha:test:finish')
-const suiteFinishCh = channel('ci:mocha:suite:finish')
-const hookErrorCh = channel('ci:mocha:hook:error')
 const parameterizedTestCh = channel('ci:mocha:test:parameterize')
 const testRunFinishCh = channel('ci:mocha:run:finish')
 
-// TODO: remove when root hooks and fixtures are implemented
-const patched = new WeakSet()
-
 function isRetry (test) {
   return test._currentRetry !== undefined && test._currentRetry !== 0
 }
 
-function getAllTestsInSuite (root) {
-  const tests = []
-  function getTests (suiteOrTest) {
-    suiteOrTest.tests.forEach(test => {
-      tests.push(test)
-    })
-    suiteOrTest.suites.forEach(suite => {
-      getTests(suite)
-    })
+function getTestAsyncResource (test) {
+  if (!test.fn) {
+    return testToAr.get(test)
   }
-  getTests(root)
-  return tests
+  if (!test.fn.asyncResource) {
+    return testToAr.get(test.fn)
+  }
+  const originalFn = originalFns.get(test.fn)
+  return testToAr.get(originalFn)
 }
 
+// TODO: remove when root hooks and fixtures are implemented
+const patched = new WeakSet()
+
+const testToAr = new WeakMap()
+const originalFns = new WeakMap()
+
 function mochaHook (Runner) {
   if (patched.has(Runner)) return Runner
 
   patched.add(Runner)
 
-  shimmer.wrap(Runner.prototype, 'runTest', runTest => function () {
-    if (!testStartCh.hasSubscribers || isRetry(this.test)) {
-      return runTest.apply(this, arguments)
+  shimmer.wrap(Runner.prototype, 'run', run => function () {
+    if (!testStartCh.hasSubscribers) {
+      return run.apply(this, arguments)
     }
 
-    const asyncResource = new AsyncResource('bound-anonymous-fn')
-    return asyncResource.runInAsyncScope(() => {
-      testStartCh.publish(this.test)
-
-      this.once('test end', AsyncResource.bind(() => {
-        let status
+    this.on('test', (test) => {
+      if (isRetry(test)) {
+        return
+      }
+      const asyncResource = new AsyncResource('bound-anonymous-fn')
+      testToAr.set(test.fn, asyncResource)
```

Comment:
we store the test's async resource here so we can run the hooks with the correct resource